cirq-core/cirq/ops/uniform_superposition_gate.py

```diff
@@ -0,0 +1,145 @@
+
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class UniformSuperpositionGate(raw_types.Gate):
+    r"""
+    Creates a generalized uniform superposition state, $\frac{1}{\sqrt{M}}\sum_{j=0}^{M-1}\ket{j}$
+    (where 1< M <= 2^n), using n qubits, according to the Shukla-Vedula algorithm [SV24].
+
+    Note: The Shukla-Vedula algorithm [SV24] offers an efficient approach for creation of a
+    generalized uniform superposition state of the form,
+    $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $, requiring only $O(\log_2 (M))$ qubits and
+    $O(\log_2 (M))$ gates. This provides an exponential improvement (in the context of reduced
+    resources and complexity) over other approaches in the literature.
+
+    Args:
+        m_value (int):
+            A positive integer M = m_value (> 1) representing the number of computational basis
+            states with an amplitude of 1/sqrt(M) in the uniform superposition state
+            ($\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $). Note that the remaining (2^n - M)
+            computational basis states have zero amplitudes. Here M need not be an integer
+            power of 2.
+
+        num_qubits (int):
+            A positive integer representing the number of qubits used.
+
+    Returns:
+        cirq.Circuit: A quantum circuit that creates the uniform superposition state:
+        $\frac{1}{\sqrt{M}} \sum_{j=0}^{M-1}  \ket{j} $.
+
+    References:
+        [SV24]
+            A. Shukla and P. Vedula, "An efficient quantum algorithm for preparation of uniform
+            quantum superposition states," Quantum Information Processing, 23(38): pp. 1-32 (2024).
+    """
+
+    def __init__(self, m_value: int, num_qubits: int) -> None:
+        """
+        Initializes UniformSuperpositionGate.
+
+        Args:
+            m_value (int): The number of computational basis states with amplitude 1/sqrt(M).
+            num_qubits (int): The number of qubits used.
+        """
+        if not (isinstance(m_value, int) and (m_value > 1)):
+            raise ValueError('m_value must be a positive integer greater than 1.')
+        if not (isinstance(num_qubits, int) and (num_qubits >= math.log2(m_value))):
+            raise ValueError(
+                'num_qubits must be an integer greater than or equal to log2(m_value).'
+            )
+        self._m_value = m_value
+        self._num_qubits = num_qubits
+        # super(UniformSuperpositionGate, self).__init__()
+
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        """
+        Decomposes the gate into a sequence of standard gates.
+
+        Args:
+            qubits (list[cirq.Qid]): Qubits to apply the gate on.
+
+        Yields:
+            cirq.Operation: Operations implementing the gate.
+        """
+        qreg = list(qubits)
+
+        if (self._m_value & (self._m_value - 1)) == 0:  # if m_value is an integer power of 2
+            m = int(math.log2(self._m_value))
```

Comment:
```suggestion
m = self._m_value.bit_length() - 1
yield H.on_each(qreg[:m])
```

Reply:
@NoureldinYosri - Thanks for the suggestion. The code changed accordingly.